ports/psoc6/machine_pdm_pcm.c

```diff
@@ -0,0 +1,664 @@
+
+    self->dma_active_buf_p = self->dma_active_buffer;
+    self->dma_processing_buf_p = self->dma_processing_buffer;
+}
+
+static inline void _dma_swap_active_dmabuf(machine_pdm_pcm_obj_t *self) {
+    uint32_t *temp = self->dma_active_buf_p;
+    self->dma_active_buf_p = self->dma_processing_buf_p;
+    self->dma_processing_buf_p = temp;
+}
+
+static void _dma_copy_from_dmabuf_to_ringbuf(machine_pdm_pcm_obj_t *self) {
+    uint8_t dma_sample_size_in_bytes = (self->bits == 16? 2 : 4) * (self->format == STEREO ? 2: 1);
+    uint8_t *dma_buff_p = (uint8_t *)self->dma_processing_buf_p;
+    uint32_t num_bytes_needed_from_ringbuf = SIZEOF_HALF_DMA_BUFFER_IN_BYTES * (PDM_PCM_RX_FRAME_SIZE_IN_BYTES / dma_sample_size_in_bytes);
+
+    // when space exists, copy samples into ring buffer
+    if (ringbuf_available_space(&self->ring_buffer) >= num_bytes_needed_from_ringbuf) {
+        uint8_t f_index = get_frame_mapping_index(self->bits, self->format);
+        uint32_t i = 0;
+        while (i < SIZEOF_HALF_DMA_BUFFER_IN_BYTES) {
+            for (uint8_t j = 0; j < PDM_PCM_RX_FRAME_SIZE_IN_BYTES; j++) {
+                int8_t r_to_a_mapping = pdm_pcm_frame_map[f_index][j];
+                if (r_to_a_mapping != -1) {
+                    ringbuf_push(&self->ring_buffer, dma_buff_p[i]);
+                    i++;
+                } else { // r_a_mapping == -1
+                    ringbuf_push(&self->ring_buffer, -1);
+                }
+            }
+        }
+    }
+}
+
+
+// =======================================================================================
+// PDM_PCM low-level abstracted functions and helper function (ports/psoc6)
+
+uint8_t pdm_pcm_get_word_byte_size(machine_pdm_pcm_obj_t *self) {
+    uint8_t res_bits = self->bits;
+    if (res_bits == 16) {
+        return 2;
+    } else {
+        return 4;
+    }
+}
+
+// Init hardware block
+static void pdm_pcm_init(machine_pdm_pcm_obj_t *self, cyhal_clock_t *clock) {
+    mplogger_print("pdm_pcm_init \r\n");
+    cyhal_pdm_pcm_cfg_t config =
+    {
+        .sample_rate = self->sample_rate,
+        .decimation_rate = self->decimation_rate,
+        .mode = self->format,
+        .word_length = self->bits,              /* bits */
+        .left_gain = self->left_gain,           /* dB */
+        .right_gain = self->right_gain,         /* dB */
+    };
+
+    cy_rslt_t result = cyhal_pdm_pcm_init(&self->pdm_pcm_obj, self->data, self->clk, &pdm_pcm_audio_clock, &config);
+    assert_pin_phy_used(result);
+    pdm_pcm_assert_raise_val("PDM_PCM initialisation failed with return code %lx !", result);
+}
+
+// Initialize audio clock
+void pdm_pcm_audio_clock_init(uint32_t audio_clock_freq_hz) {
+    mplogger_print("pdm_pcm_audio_clock_init \r\n");
+    cyhal_clock_t pll_clock;
+    cy_rslt_t result;
+
+    static bool clock_set = false;
+
+    result = cyhal_clock_reserve(&pll_clock, &CYHAL_CLOCK_PLL[0]);
```

Comment:
Now that the clock configuration is set externally we should refactor this or?

Reply:
Yep..true! I take that in test ticket.